subsys/mgmt/mcumgr/smp/src/client.c

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** SMP - Simple Management Client Protocol. */
+
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/net/buf.h>
+#include <zephyr/mgmt/mcumgr/mgmt/mgmt.h>
+#include <zephyr/mgmt/mcumgr/smp/smp.h>
+#include <zephyr/mgmt/mcumgr/smp/smp_client.h>
+#include <zephyr/mgmt/mcumgr/transport/smp.h>
+#include <mgmt/mcumgr/transport/smp_internal.h>
+#include <assert.h>
+#include <string.h>
+
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+
+#include <zephyr/logging/log.h>
+
+#define LOG_MODULE_NAME smp_client
+LOG_MODULE_REGISTER(smp_client);
+
+#define SMP_CMD_MAX 4
+
+struct smp_client_cmd_req {
+	sys_snode_t node;
+	struct net_buf *nb;
+	void *user_data;
+	smp_client_res_fn cb;
+	int64_t timestamp;
+	int retry_cnt;
+};
+
+struct smp_client_data_base {
+	struct k_work_delayable work_delay;
+	sys_slist_t cmd_free_list;
+	sys_slist_t cmd_list;
+	/* SMP SEQ */
+	uint8_t smp_seq;
+};
+
+static struct smp_client_cmd_req smp_cmd_req_bufs[SMP_CMD_MAX];
+static struct smp_client_data_base smp_client_data;
```

Comment:
A single client, and with global file scope? That does not seem very extensible.